fsspec/spec.py

```diff
@@ -428,11 +428,11 @@ def walk(self, path, maxdepth=None, topdown=True, on_error="omit", **kwargs):
         except (FileNotFoundError, OSError) as e:
             if on_error == "raise":
                 raise
-            elif callable(on_error):
+            if callable(on_error):
                 on_error(e)
-            if detail:
-                return path, {}, {}
-            return path, [], []
+            result = (path, {}, {}) if detail else (path, [], [])
+            yield result
+            return result
```

Comment:
You can't both yield and return the result, else there will be two tuples produced, not just once. I think you are right that the non-return variant of this is
```
yield result
raise StopIteration
```
which to my mind is far less attractive than `return result`.